pkg/keyspace/tso_keyspace_group.go

```diff
@@ -16,24 +16,56 @@ package keyspace
 
 import (
 	"context"
+	"strconv"
+	"sync"
 
+	"github.com/tikv/pd/pkg/balancer"
+	"github.com/tikv/pd/pkg/mcs/discovery"
 	"github.com/tikv/pd/pkg/mcs/utils"
 	"github.com/tikv/pd/pkg/storage/endpoint"
 	"github.com/tikv/pd/pkg/storage/kv"
+	"github.com/tikv/pd/pkg/utils/etcdutil"
+	"go.etcd.io/etcd/clientv3"
+)
+
+const (
+	defaultBalancerPolicy = balancer.PolicyRoundRobin
 )
 
 // GroupManager is the manager of keyspace group related data.
 type GroupManager struct {
 	ctx context.Context
 	// store is the storage for keyspace group related information.
 	store endpoint.KeyspaceGroupStorage
+
+	client *clientv3.Client
+
+	clusterID uint64
+
+	policy balancer.Policy
+
+	nodesBalancer balancer.Balancer[string]
+
+	groups struct {
+		sync.RWMutex
+		balancerMap map[string]balancer.Balancer[uint32]
+	}
 }
 
 // NewKeyspaceGroupManager creates a Manager of keyspace group related data.
-func NewKeyspaceGroupManager(ctx context.Context, store endpoint.KeyspaceGroupStorage) *GroupManager {
+func NewKeyspaceGroupManager(ctx context.Context, store endpoint.KeyspaceGroupStorage, client *clientv3.Client, clusterID uint64) *GroupManager {
 	return &GroupManager{
-		ctx:   ctx,
-		store: store,
+		ctx:       ctx,
+		store:     store,
+		client:    client,
+		clusterID: clusterID,
+		policy:    defaultBalancerPolicy,
+		groups: struct {
```

Comment:
The name is confusing since it is not group level but user kind level.

Reply:
maybe. It has been removed.